pkg/workloadlearning/cache.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workloadlearning
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// ReadTableCostCache stores the cached workload learning metrics
+type ReadTableCostCache struct {
+	TableCostMetrics map[int64]*ReadTableCostMetrics // key: tableID
+	Version          uint64
+}
+
+// WLCacheWorker the worker to cache all workload-related metrics
+// Now it is also used to save the cache data of table cost metrics.
+type WLCacheWorker struct {
+	sysSessionPool     util.SessionPool
+	readTableCostCache *ReadTableCostCache
+	sync.RWMutex
+}
+
+// NewWLCacheWorker Create a new workload learning cache worker to cache all workload-related metrics
+// from storage mysql.tidb_workload_values to memory
+func NewWLCacheWorker(pool util.SessionPool) *WLCacheWorker {
+	return &WLCacheWorker{pool, &ReadTableCostCache{}, sync.RWMutex{}}
+}
+
+// UpdateTableCostCache refreshes the cached workload learning metrics
+func (cw *WLCacheWorker) UpdateTableCostCache() error {
+	// Get latest metrics from storage
+	se, err := cw.sysSessionPool.Get()
+	if err != nil {
+		return err
+	}
+	defer cw.sysSessionPool.Put(se)
```

Comment:
Please destroy this session instead of putting it back to avoid data corruption.